middle_end/flambda/simplify/simplify_set_of_closures.ml

```diff
@@ -457,25 +456,36 @@ let simplify_function context ~used_closure_vars ~shareable_constants
       Code_id.Map.find old_code_id (C.old_to_new_code_ids_all_sets context)
     in
     let code =
-      code
-      |> Code.with_code_id new_code_id
-      |> Code.with_newer_version_of (Some old_code_id)
-      |> Code.with_params_and_body
-           ~cost_metrics
-           (Present (params_and_body, free_names_of_code))
-    in
-    let code =
-      Rebuilt_static_const.create (Code code) ~free_names:Unknown
+      Rebuilt_static_const.create_code
+        (DA.are_rebuilding_terms dacc_after_body)
+        new_code_id
+        ~params_and_body:(Present (params_and_body, free_names_of_code))
+        ~newer_version_of:(Some old_code_id)
+        ~params_arity:(Code.params_arity code)
+        ~result_arity:(Code.result_arity code)
+        ~stub:(Code.stub code)
+        ~inline:(Code.inline code)
+        ~is_a_functor:(Code.is_a_functor code)
+        ~recursive:(Code.recursive code)
+        ~cost_metrics
     in
     let function_decl = FD.update_code_id function_decl new_code_id in
     let function_type =
-      (* We need to manually specify the cost metrics to use to ensure that
-         they are the one of the body after simplification. *)
-      function_decl_type
-        ~pass:Inlining_report.After_simplify
-        ~cost_metrics_source:(Metrics cost_metrics)
-        (DA.denv dacc_after_body) function_decl
-        Rec_info.initial
+      (* When not rebuilding terms we cannot infer function types.  However
```

Comment:
What about enforcing it by setting inlining to be off when we are not rebuilding terms. Or at least emitting a warning if that's the case. It should never happen in practice but we can't be sure that the assumption will still hold in the future.

Reply:
I'm not sure we need to actually enforce it, but inlining is disabled whenever `DE.set_do_not_rebuild_terms` is called.

Reply:
Wouldn't that mean that when doing speculative inlining, we do not consider the potential benefit of inlining a function inside the body of the function which we're trying to speculatively inline ?
For instance imagine the following code:
```
let f b x =
  if b then x else ... <some very big code>
  
let g x = function
  | A -> f true x
  | _ -> <some more big code>
  
let h x =
  g A x
```

I'd expect that we'd need to speculatively inline `g` and in that speculation also inline `f`, for it to be beneficial, but I feel that disabling inlining when doing speculative inlining would prevent that. Did i miss something ?

Reply:
That is correct, but also intended. We don't think that the cost of recursively doing speculation, both in terms of compile times and in terms of complexity of the inliner are worth it just to inline cases like those.

(FWIW I would actually expect your exact example to get inlined anyway, since inlining `g` on its own removes a branch and doesn't increase the code size).

Reply:
@Gbury Please follow up with @lpw25 e.g. on Slack if further discussion is needed!